pkg/controller/pod.go

```diff
@@ -192,25 +204,106 @@ func (c *Controller) syncPod(ctx context.Context, pod *corev1.Pod) error {
 		return nil
 	}
 
+	if virtualNode.Data.Spec.ServiceDiscovery.AwsCloudMap.NamespaceName == nil ||
+		virtualNode.Data.Spec.ServiceDiscovery.AwsCloudMap.ServiceName == nil {
+		klog.Errorf("CloudMap NamespaceName or ServiceName is null")
+		return nil
+	}
+
 	cloudmapConfig := virtualNode.Data.Spec.ServiceDiscovery.AwsCloudMap
 
+	var serviceItem *cloudMapInstanceCacheItem
+	serviceInstanceSummary := make(map[string]bool)
+	cloudMapServiceKey := *cloudmapConfig.NamespaceName + "-" + *cloudmapConfig.ServiceName
+
+	existingItem, exists, _ := c.cloudMapInstanceCache.Get(&cloudMapInstanceCacheItem{
+		key: cloudMapServiceKey,
+	})
+
 	if !pod.DeletionTimestamp.IsZero() {
 		klog.V(4).Infof("Deregistering instance %s under service %+v", pod.Name, cloudmapConfig)
 		err = c.cloud.DeregisterInstance(ctx, instanceID, cloudmapConfig)
 		if err != nil {
 			return err
 		}
+		//Clear the instance from Cache.
+		if exists {
+			delete(existingItem.(*cloudMapInstanceCacheItem).instanceSummary, pod.Status.PodIP)
+		}
+		return nil
+	}
+
+	if exists {
+		serviceItem = existingItem.(*cloudMapInstanceCacheItem)
+	} else {
+		//Retrieve CloudMap instances for this service
+		appmeshCloudMapConfig := &appmesh.AwsCloudMapServiceDiscovery{
+			NamespaceName: cloudmapConfig.NamespaceName,
+			ServiceName:   cloudmapConfig.ServiceName,
+		}
+
+		if serviceInstances, _ := c.getServiceInstancesFromCloudMap(ctx, appmeshCloudMapConfig); len(serviceInstances) > 0 {
```

Comment:
`if serviceInstances, _ := c.getServiceInstancesFromCloudMap` 

a general rule in golang is check for error, and avoid use any other result if error != nil. e.g. a bad implementation might return a partially populated `serviceInstances` with an error.